src/persist/src/indexed/encoding.rs

```diff
@@ -0,0 +1,469 @@
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<&Description<SeqNo>> = None;
+        for (desc, _) in self.batches.iter() {
+            if let Some(prev) = prev {
+                if prev.upper() != desc.lower() {
+                    return Err(format!(
+                        "invalid batch sequence: {:?} followed by {:?}",
+                        prev, desc
+                    )
+                    .into());
+                }
+            }
+            prev = Some(desc)
+        }
+        Ok(())
+    }
+}
+
+impl BlobTraceMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<&Description<u64>> = None;
+        for (desc, _) in self.batches.iter() {
+            if let Some(prev) = prev {
+                if prev.upper() != desc.lower() {
+                    return Err(format!(
+                        "invalid batch sequence: {:?} followed by {:?}",
+                        prev, desc,
+                    )
+                    .into());
+                }
+            }
+            prev = Some(desc)
+        }
+        Ok(())
+    }
+}
+
+impl BlobFutureBatch {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<(&u64, &String, &String)> = None;
+        for (seqno, (key, val), ts, _) in self.updates.iter() {
+            // Check seqno against desc.
+            if !self.desc.lower().less_equal(seqno) {
+                return Err(format!(
+                    "seqno {:?} is before the batch lower: {:?}",
+                    seqno, self.desc
+                )
+                .into());
+            }
+            if self.desc.upper().less_equal(seqno) {
+                return Err(format!(
+                    "seqno {:?} is after or equal to the batch upper: {:?}",
+                    seqno, self.desc
+                )
+                .into());
+            }
+
+            // Check ordering.
+            let this = (ts, key, val);
+            if let Some(prev) = prev {
+                match prev.cmp(&this) {
+                    Ordering::Less => {} // Correct.
+                    Ordering::Equal => return Err(format!("unconsolidated: {:?}", this).into()),
+                    Ordering::Greater => {
+                        return Err(format!("unsorted: {:?} was before {:?}", prev, this).into())
+                    }
+                }
+            }
+            prev = Some(this);
+        }
+
+        Ok(())
+    }
+}
+
+impl BlobTraceBatch {
+    /// Asserts the documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<(&String, &String, &u64)> = None;
+        for ((key, val), ts, _) in self.updates.iter() {
```

Comment:
should we assert that the diff is nonzero

Reply:
Great idea!